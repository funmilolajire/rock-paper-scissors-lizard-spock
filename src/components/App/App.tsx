import '../../sass/main.scss'

export const App = () => {
  return (
    <div className="App">
    </div>
  );
}