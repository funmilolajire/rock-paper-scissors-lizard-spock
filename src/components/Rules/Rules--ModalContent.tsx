import { IoMdClose } from 'react-icons/io';
import rules from './images/rules.svg';

export const ModalContent = () => {
    return (
        <div className="ModalContent">
            <section className="top">
                <h3>Rules</h3>
                <span className="close"><IoMdClose /></span>
            </section>
            <figure>
                <img src={rules} alt="rules" />
            </figure>
        </div>
    )
}