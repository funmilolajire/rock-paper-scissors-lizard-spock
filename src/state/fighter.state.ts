import { createState, useState, State, DevTools } from '@hookstate/core';
import { outcome } from './utils/rules';
import { DevToolsInitialize } from '@hookstate/devtools';

export const fighters = ["rock", "paper", "scissors", "lizard", "spock"];

const playerFighter = createState('');
const houseFighter = createState('');

const wrapState = (state: State<string>) => ({
    get: () => state.value,
    select: (name: string) => state.set(prev => prev = name),
})

//to use outside of a component
export const accessPlayerFighter = () => wrapState(playerFighter);
export const accessHouseFighter = () => wrapState(houseFighter);

//to use within a component
export const usePlayerFighter = () => wrapState(useState(playerFighter));
export const useHouseFighter = () => wrapState(useState(houseFighter));

//devtools
DevTools(playerFighter).label('player-fighter')
DevTools(houseFighter).label('house-fighter')
DevToolsInitialize({
    monitored: ['player-fighter', 'house-fighter'],
    callstacksDepth: 30
})
