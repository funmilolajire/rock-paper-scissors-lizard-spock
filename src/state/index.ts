export * from './modal.state';
export * from './score.state';
export * from './fighter.state';