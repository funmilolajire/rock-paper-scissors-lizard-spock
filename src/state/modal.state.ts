import { createState, useState, State, DevTools } from '@hookstate/core';
import { DevToolsInitialize } from '@hookstate/devtools';

const modalState = createState(false);
const wrapState = (state: State<boolean>) => ({
    get: () => state.value,
    open: () => state.set(true),
    close: () => state.set(false)
})

//to use outside of a component
export const accessModalState = () => wrapState(modalState);

//to use within a component
export const useModalState = () => wrapState(useState(modalState));

//devtools
DevTools(modalState).label('modal-state')
DevToolsInitialize({
    monitored: ['modal-state'],
    callstacksDepth: 30
})
