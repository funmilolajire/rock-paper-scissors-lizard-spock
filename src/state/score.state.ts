import { createState, useState, State, DevTools } from '@hookstate/core';
import { DevToolsInitialize } from '@hookstate/devtools';
import { Persistence } from '@hookstate/persistence'


const playerState = createState(0);
const houseState = createState(0);

const wrapState = (state: State<number>) => ({
    get: () => state.value,
    increment: () => state.set(prev => prev + 1),
    clear: () => state.set(0)
})

//to use from outside a component
export const accessPlayerState = () => wrapState(playerState.attach(Persistence('playerScore')))
export const accessHouseState = () => wrapState(houseState.attach(Persistence('houseScore')))

//to use from within a component
export const usePlayerState = () => wrapState(useState(playerState).attach(Persistence('playerScore')))
export const useHouseState = () => wrapState(useState(houseState).attach(Persistence('houseScore')))

//devtools
DevTools(playerState).label('player-state')
DevTools(houseState).label('house-state')
DevToolsInitialize({
    monitored: ['player-state', 'house-state'],
    callstacksDepth: 30
})
