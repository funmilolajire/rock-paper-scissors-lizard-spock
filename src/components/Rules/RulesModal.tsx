import ReactDOM from 'react-dom';
import { ModalContent } from './Rules--ModalContent';

export const RulesModal = () => {
    return ReactDOM.createPortal(
        <div className="RulesModal">
            <ModalContent />
        </div>
        , document.getElementById('portal') as HTMLElement)
}