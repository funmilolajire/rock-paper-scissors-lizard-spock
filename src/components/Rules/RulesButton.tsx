import { useModalState } from '../../state';
export const RulesButton = () => {
    //open Modal on click
    const modalState = useModalState();
    const openModal = () => modalState.open();

    return (
        <button onClick={openModal} className="RulesButton">
            Rules
        </button>
    )
}