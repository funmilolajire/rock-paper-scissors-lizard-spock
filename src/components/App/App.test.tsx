import { mount } from 'enzyme';
import { act } from '@testing-library/react';
import { App } from './App';
import { RulesButton } from '../Rules/RulesButton';
import { RulesModal } from '../Rules/RulesModal';
import { accessModalState } from '../../state';

describe('app component', () => {
  let component = mount(<App />)
  test('make a snapshot of app component', () => {
    expect(component).toMatchSnapshot()
  })
  test('check that it renders one initial component', () => {
    const getComponents = component.containsAnyMatchingElements([<RulesButton />])
    expect(getComponents).toBeTrue();
  })
  test('check that it doesn\'t render modal on load', () => {
    const getComponents = component.containsAnyMatchingElements([<RulesModal />])
    expect(getComponents).toBeFalse();
  })
  test('check that it renders modal on click', () => {
    const portal = global.document.createElement('div')
    portal.setAttribute('id', 'portal')
    const body = global.document.querySelector('body')
    body?.appendChild(portal)
    act(() => { accessModalState().open() })
    expect(accessModalState().get()).toBeTrue();
  })
})