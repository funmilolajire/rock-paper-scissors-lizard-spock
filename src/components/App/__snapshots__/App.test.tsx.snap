// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app component make a snapshot of app component 1`] = `
<App>
  <div
    className="App"
  >
    <Header>
      <header
        className="Header"
      >
        <div
          className="Header--Inner"
        >
          <NameList>
            <ul
              className="NameList"
            >
              <li>
                Rock
              </li>
              <li>
                Paper
              </li>
              <li>
                Scissors
              </li>
              <li>
                Lizard
              </li>
              <li>
                Spock
              </li>
            </ul>
          </NameList>
          <Score>
            <section
              className="Score"
            >
              <h2>
                Score
              </h2>
              <span>
                0
                 - 
                0
              </span>
            </section>
          </Score>
        </div>
      </header>
    </Header>
    <Main>
      <div
        className="Main"
      >
        <GameLoad>
          <div
            className="GameLoad"
          >
            <Polygon>
              <div
                className="polygon--container"
              >
                <div
                  className="outer--polygon"
                >
                  <div
                    className="inner--polygon"
                  >
                    Yellow
                  </div>
                </div>
              </div>
            </Polygon>
            <Icon
              iconName="rock"
              key="rock"
              onClick={[Function]}
            >
              <div
                className="Icon rock"
                data-icon-name="rock"
                onClick={[Function]}
                style={
                  Object {
                    "background": "#9D1634",
                  }
                }
              >
                <div
                  className="Icon--OuterRing"
                  style={
                    Object {
                      "background": "#DB2E4D",
                    }
                  }
                >
                  <div
                    className="Icon--InnerRing"
                  >
                    <div
                      className="Icon--InnerInnerRing"
                    >
                      <img
                        alt="rock"
                        src="rock.svg"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </Icon>
            <Icon
              iconName="paper"
              key="paper"
              onClick={[Function]}
            >
              <div
                className="Icon paper"
                data-icon-name="paper"
                onClick={[Function]}
                style={
                  Object {
                    "background": "#2A45C2",
                  }
                }
              >
                <div
                  className="Icon--OuterRing"
                  style={
                    Object {
                      "background": "#4664F4",
                    }
                  }
                >
                  <div
                    className="Icon--InnerRing"
                  >
                    <div
                      className="Icon--InnerInnerRing"
                    >
                      <img
                        alt="paper"
                        src="paper.svg"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </Icon>
            <Icon
              iconName="scissors"
              key="scissors"
              onClick={[Function]}
            >
              <div
                className="Icon scissors"
                data-icon-name="scissors"
                onClick={[Function]}
                style={
                  Object {
                    "background": "#C76C1B",
                  }
                }
              >
                <div
                  className="Icon--OuterRing"
                  style={
                    Object {
                      "background": "#EB9F0E",
                    }
                  }
                >
                  <div
                    className="Icon--InnerRing"
                  >
                    <div
                      className="Icon--InnerInnerRing"
                    >
                      <img
                        alt="scissors"
                        src="scissors.svg"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </Icon>
            <Icon
              iconName="lizard"
              key="lizard"
              onClick={[Function]}
            >
              <div
                className="Icon lizard"
                data-icon-name="lizard"
                onClick={[Function]}
                style={
                  Object {
                    "background": "#5F37A8",
                  }
                }
              >
                <div
                  className="Icon--OuterRing"
                  style={
                    Object {
                      "background": "#834EE3",
                    }
                  }
                >
                  <div
                    className="Icon--InnerRing"
                  >
                    <div
                      className="Icon--InnerInnerRing"
                    >
                      <img
                        alt="lizard"
                        src="lizard.svg"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </Icon>
            <Icon
              iconName="spock"
              key="spock"
              onClick={[Function]}
            >
              <div
                className="Icon spock"
                data-icon-name="spock"
                onClick={[Function]}
                style={
                  Object {
                    "background": "#2D8DAB",
                  }
                }
              >
                <div
                  className="Icon--OuterRing"
                  style={
                    Object {
                      "background": "#3FB7CD",
                    }
                  }
                >
                  <div
                    className="Icon--InnerRing"
                  >
                    <div
                      className="Icon--InnerInnerRing"
                    >
                      <img
                        alt="spock"
                        src="spock.svg"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </Icon>
          </div>
        </GameLoad>
      </div>
    </Main>
    <RulesButton>
      <button
        className="RulesButton"
        onClick={[Function]}
      >
        Rules
      </button>
    </RulesButton>
  </div>
</App>
`;
